### PR TITLE
add available device to regression tests #3335

### PR DESCRIPTION
**Pull request by BanzaiTokyo**

### CODE REVIEW

File: tests/ignite/metrics/regression/test_canberra_metric.py [hunk 1]
```diff
@@ -58,45 +59,48 @@ def test_compute():
...
+    data = list(range(y_pred.shape[0] // batch_size))
+    cm = engine.run(data, max_epochs=1).metrics["cm"]
 
-    for _ in range(5):
-        # check multiple random inputs as random exact occurencies are rare
-        test_cases = get_test_cases()
-        for y_pred, y, batch_size in test_cases:
-            _test(y_pred, y, batch_size)
+    X = y_pred.cpu().numpy().reshape(len(y_pred), -1)
+    Y = y.cpu().numpy().reshape(len(y), -1)
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_canberra_metric.py:96]: Let's avoid variables in upper case, lint wont pass most probably.
Also, x and y for y_pred and y are not appropriate names.

File: tests/ignite/metrics/regression/test_canberra_metric.py [hunk 2]
```diff
@@ -58,45 +59,48 @@ def test_compute():
...
 
-        canberra = DistanceMetric.get_metric("canberra")
+    m = CanberraMetric(device=available_device)
+    print(f"m's dtype: {m._double_dtype}")
+    assert m._device == torch.device(available_device)
 
-        data = list(range(y_pred.shape[0] // batch_size))
-        cm = engine.run(data, max_epochs=1).metrics["cm"]
+    m.attach(engine, "cm")
+    print(f"m's dtype again: {m._double_dtype}")
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_canberra_metric.py:88]: Remove all prints

File: tests/ignite/metrics/regression/test_canberra_metric.py [hunk 3]
```diff
@@ -58,45 +59,48 @@ def test_compute():
...
-        engine = Engine(update_fn)
+@pytest.mark.parametrize("n_times", range(5))
+@pytest.mark.parametrize(
+    "test_cases",
+    [
+        (torch.rand(size=(100,)), torch.rand(size=(100,)), 10),
+        (torch.rand(size=(100, 1)), torch.rand(size=(100, 1)), 20),
+    ],
+)
+def test_integration(n_times, test_cases, available_device):
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_canberra_metric.py:70]: Good work with rewriting the test

File: tests/ignite/metrics/regression/test_canberra_metric.py [hunk 4]
```diff
@@ -58,45 +59,48 @@ def test_compute():
...
-def test_integration():
-    def _test(y_pred, y, batch_size):
-        def update_fn(engine, batch):
-            idx = (engine.state.iteration - 1) * batch_size
-            y_true_batch = np_y[idx : idx + batch_size]
-            y_pred_batch = np_y_pred[idx : idx + batch_size]
-            return torch.from_numpy(y_pred_batch), torch.from_numpy(y_true_batch)
-
-        engine = Engine(update_fn)
+@pytest.mark.parametrize("n_times", range(5))
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_canberra_metric.py:62]:
```suggestion
@pytest.mark.parametrize("n_times", range(3))
```

File: tests/ignite/metrics/regression/test_fractional_absolute_error.py [hunk 5]
```diff
@@ -28,14 +28,15 @@ def test_wrong_input_shapes():
...
-def test_compute():
+def test_compute(available_device):
     a = np.random.randn(4)
     b = np.random.randn(4)
     c = np.random.randn(4)
     d = np.random.randn(4)
     ground_truth = np.random.randn(4)
 
-    m = FractionalAbsoluteError()
+    m = FractionalAbsoluteError(device=available_device)
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_fractional_absolute_error.py:38]: Let's rewrite this test: let's create a, b, c, d variables with torch tensor API instead of numpy

File: tests/ignite/metrics/regression/test_fractional_bias.py [hunk 6]
```diff
@@ -28,14 +28,15 @@ def test_wrong_input_shapes():
...
-def test_fractional_bias():
+def test_fractional_bias(available_device):
     a = np.random.randn(4)
     b = np.random.randn(4)
     c = np.random.randn(4)
     d = np.random.randn(4)
     ground_truth = np.random.randn(4)
 
-    m = FractionalBias()
+    m = FractionalBias(device=available_device)
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_fractional_bias.py:38]: Same here: a, b, c, d with torch tensor API

File: tests/ignite/metrics/test_epoch_metric.py [hunk 7]
```diff
@@ -103,7 +103,7 @@ def compute_fn(y_preds, y_targets):
     targets = torch.cat([output1[1], output2[1], output3[1]], dim=0)
 
     result = em.compute()
-    assert result == compute_fn(preds, targets)
+    assert result == pytest.approx(compute_fn(preds, targets), rel=1e-6)
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:106]: This change is surprising, why do you need this?

**Review by vfdev-5 — COMMENTED** [review]
I left few comments. Now this PR starts being too large, let's split it and work chunk by chunk.

File: tests/ignite/metrics/test_epoch_metric.py (hunk 7 above)
Comment (BanzaiTokyo) [reply tests/ignite/metrics/test_epoch_metric.py:106]: because it was not exactly the same and the comparison failed

File: ignite/metrics/regression/_base.py [hunk 8]
```diff
@@ -30,6 +30,9 @@ def _check_output_types(output: Tuple[torch.Tensor, torch.Tensor]) -> None:
 
 
 def _torch_median(output: torch.Tensor) -> float:
+    if output.device.type == "mps":
```
[the remark this answers is not in the document]
Comment (BanzaiTokyo) [reply ignite/metrics/regression/_base.py:33]: torch.kthvalue is not supported on MPS. adding comment

File: tests/ignite/metrics/test_epoch_metric.py (hunk 7 above)
Comment (vfdev-5) [reply tests/ignite/metrics/test_epoch_metric.py:106]: On MPS ?

Comment (vfdev-5) [reply tests/ignite/metrics/test_epoch_metric.py:106]: Make a separate PR for this metric

File: ignite/metrics/regression/canberra_metric.py [hunk 9]
```diff
@@ -73,8 +73,9 @@ def reset(self) -> None:
 
     def _update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
         y_pred, y = output[0].detach(), output[1].detach()
-        errors = torch.abs(y - y_pred) / (torch.abs(y_pred) + torch.abs(y) + 1e-15)
-        self._sum_of_errors += torch.sum(errors).to(self._device)
+        eps = torch.tensor(1e-15, dtype=self._double_dtype, device=y.device)
```
[the remark this answers is not in the document]
Comment (BanzaiTokyo) [reply ignite/metrics/regression/canberra_metric.py:76]: I believe unwrapped 1e-15 may accidentally cast errors to f64

[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/metrics/regression/canberra_metric.py:76]: You should check your assumption.

[the remark this answers is not in the document]
Comment (BanzaiTokyo) [reply ignite/metrics/regression/canberra_metric.py]: you are right! switched back

File: ignite/metrics/regression/fractional_bias.py [hunk 10]
```diff
@@ -64,13 +64,13 @@ class FractionalBias(_BaseRegression):
 
     @reinit__is_reduced
     def reset(self) -> None:
-        self._sum_of_errors = torch.tensor(0.0, dtype=torch.double, device=self._device)
+        self._sum_of_errors = torch.tensor(0.0, dtype=self._double_dtype, device=self._device)
```
[the remark this answers is not in the document]
Comment (BanzaiTokyo) [reply ignite/metrics/regression/fractional_bias.py:67]: by everywhere you mean where exactly?

[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/metrics/regression/fractional_bias.py:67]: In all regression metrics

File: ignite/metrics/regression/mean_absolute_relative_error.py [hunk 11]
```diff
@@ -72,7 +72,9 @@ def _update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
         if (y == 0).any():
             raise NotComputableError("The ground truth has 0.")
         absolute_error = torch.abs(y_pred - y.view_as(y_pred)) / torch.abs(y.view_as(y_pred))
-        self._sum_of_absolute_relative_errors += torch.sum(absolute_error).to(self._device)
+        self._sum_of_absolute_relative_errors += torch.sum(absolute_error).to(
```
Comment (vfdev-5) [inline ignite/metrics/regression/mean_absolute_relative_error.py:75]: Let's update reset() method as in other implementation to use double precision

File: ignite/metrics/regression/pearson_correlation.py [hunk 12]
```diff
@@ -87,11 +87,11 @@ def reset(self) -> None:
 
     def _update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
         y_pred, y = output[0].detach(), output[1].detach()
-        self._sum_of_y_preds += y_pred.sum().to(self._device)
-        self._sum_of_ys += y.sum().to(self._device)
-        self._sum_of_y_pred_squares += y_pred.square().sum().to(self._device)
-        self._sum_of_y_squares += y.square().sum().to(self._device)
-        self._sum_of_products += (y_pred * y).sum().to(self._device)
+        self._sum_of_y_preds += y_pred.sum().to(dtype=self._double_dtype, device=self._device)
```
Comment (vfdev-5) [inline ignite/metrics/regression/pearson_correlation.py:90]: Same here, check and update reset() method

File: tests/ignite/metrics/regression/test_canberra_metric.py [hunk 13]
```diff
@@ -20,14 +20,15 @@ def test_wrong_input_shapes():
...
-def test_compute():
+def test_compute(available_device):
     a = np.random.randn(4)
     b = np.random.randn(4)
     c = np.random.randn(4)
     d = np.random.randn(4)
     ground_truth = np.random.randn(4)
 
-    m = CanberraMetric()
+    m = CanberraMetric(device=available_device)
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_canberra_metric.py:30]: Please replace a, b, c, d creation from numpy to torch API

**Comment by BanzaiTokyo** [comment]
@vfdev-5 I believe I have addressed your comments, please, review again to make sure everything is ok

File: ignite/metrics/regression/pearson_correlation.py [hunk 14]
```diff
@@ -119,5 +119,12 @@ def compute(self) -> float:
...
         y_var = torch.clamp(y_var, min=0.0)
 
-        r = cov / torch.clamp(torch.sqrt(y_pred_var * y_var), min=self.eps)
+        denom = y_pred_var * y_var
+        denom = torch.clamp(denom, min=self.eps)
+        denom = torch.sqrt(denom)
+        r = cov / denom
+
+        if torch.isnan(r):
+            return 0.0
```
Comment (vfdev-5) [inline ignite/metrics/regression/pearson_correlation.py:128]: Why this change ? Have you checked with the original implementation?

File: ignite/metrics/regression/fractional_bias.py [hunk 15]
```diff
@@ -64,7 +64,7 @@ class FractionalBias(_BaseRegression):
 
     @reinit__is_reduced
     def reset(self) -> None:
-        self._sum_of_errors = torch.tensor(0.0, dtype=torch.double, device=self._device)
+        self._sum_of_errors = torch.tensor(0.0, device=self._device)
```
Comment (vfdev-5) [inline ignite/metrics/regression/fractional_bias.py:67]: Keep double precision if it was already done here, just replace with `self._double_dtype`

File: tests/ignite/metrics/regression/test_canberra_metric.py [hunk 16]
```diff
@@ -20,83 +20,85 @@ def test_wrong_input_shapes():
...
+    "test_cases",
+    [
+        (torch.rand(size=(100,)), torch.rand(size=(100,)), 10),
+        (torch.rand(size=(100, 1)), torch.rand(size=(100, 1)), 20),
+    ],
+)
+def test_integration(n_times, test_cases, available_device):
+    y_pred, y, batch_size = test_cases
+    assert y_pred.dtype == torch.float32
+    assert y.dtype == torch.float32
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_canberra_metric.py:73]:
```suggestion
```

File: tests/ignite/metrics/regression/test_canberra_metric.py [hunk 17]
```diff
@@ -20,83 +20,85 @@ def test_wrong_input_shapes():
...
 
-        m = CanberraMetric()
-        m.attach(engine, "cm")
+    def update_fn(engine, batch):
+        idx = (engine.state.iteration - 1) * batch_size
+        y_true_batch = y[idx : idx + batch_size].to(dtype=torch.float32)
+        y_pred_batch = y_pred[idx : idx + batch_size].to(dtype=torch.float32)
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_canberra_metric.py:78]:
```suggestion
        y_true_batch = y[idx : idx + batch_size]
        y_pred_batch = y_pred[idx : idx + batch_size]
```

File: tests/ignite/metrics/regression/test_fractional_absolute_error.py [hunk 18]
```diff
@@ -98,7 +90,7 @@ def get_test_cases():
         # check multiple random inputs as random exact occurencies are rare
         test_cases = get_test_cases()
         for y_pred, y, batch_size in test_cases:
-            _test(y_pred, y, batch_size)
+            _test(y_pred, y, batch_size, device=available_device)
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_fractional_absolute_error.py:93]: Please rewrite this test similarly as you have done in tests/ignite/metrics/regression/test_canberra_metric.py, test_integration

File: tests/ignite/metrics/regression/test_fractional_bias.py [hunk 19]
```diff
@@ -28,51 +27,53 @@ def test_wrong_input_shapes():
...
-    np_sum += (2 * (ground_truth - d) / (d + ground_truth)).sum()
-    np_len += len(d)
-    np_ans = np_sum / np_len
-    assert m.compute() == pytest.approx(np_ans)
-
-
-def test_integration():
-    def _test(y_pred, y, batch_size):
+
+def test_integration(available_device):
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_fractional_bias.py:54]: Same here, let's rewrite this test as other test_integration

File: tests/ignite/metrics/regression/test_geometric_mean_absolute_error.py [hunk 20]
```diff
@@ -28,47 +28,57 @@ def test_wrong_input_shapes():
...
     a = np.random.randn(4)
     b = np.random.randn(4)
     c = np.random.randn(4)
     d = np.random.randn(4)
     ground_truth = np.random.randn(4)
     np_prod = 1.0
 
-    m = GeometricMeanAbsoluteError()
-    m.update((torch.from_numpy(a), torch.from_numpy(ground_truth)))
+    m = GeometricMeanAbsoluteError(device=available_device)
```
Comment (vfdev-5) [inline tests/ignite/metrics/regression/test_geometric_mean_absolute_error.py:39]: Define a, b, c, d, ground_truth using torch API

**Review by vfdev-5 — COMMENTED** [review]
Let's split this PR into multiple ones and make them merge once you are done porting 1 or 2 metrics. We want to do the following:
- add available_device
- use tensor API instead of numpy for data creation
- rewrite test_integration tests using pytest parametrize instead of nested _test methods

File: ignite/metrics/regression/pearson_correlation.py (hunk 14 above)
Comment (BanzaiTokyo) [reply ignite/metrics/regression/pearson_correlation.py:128]: just to make sure I ran the tests once again. Without this snippet, on MPS a division by zero produces NaN:
```
FAILED tests/ignite/metrics/regression/test_pearson_correlation.py::test_degenerated_sample[mps] - assert 0.0 ± 1.0e-12 == nan
  
  comparison failed
  Obtained: nan
  Expected: 0.0 ± 1.0e-12
```
